extension/api.js

```diff
@@ -0,0 +1,233 @@
+          context,
+          name: "addonsSearchExperiment.onSearchEngineModified",
+          register: (fire) => {
+            const onSearchEngineModifiedObserver = (
+              aSubject,
+              aTopic,
+              aData
+            ) => {
+              if (
+                aTopic !== SEARCH_TOPIC_ENGINE_MODIFIED ||
+                // We are only interested in these modified types.
+                !["engine-added", "engine-removed", "engine-changed"].includes(
+                  aData
+                )
+              ) {
+                return;
+              }
+
+              fire.async();
+            };
+
+            Services.obs.addObserver(
+              onSearchEngineModifiedObserver,
+              SEARCH_TOPIC_ENGINE_MODIFIED
+            );
+
+            return () => {
+              Services.obs.removeObserver(
+                onSearchEngineModifiedObserver,
+                SEARCH_TOPIC_ENGINE_MODIFIED
+              );
+            };
+          },
+        }).api(),
+
+        // `onRedirected` is an event fired after a request has stopped and
+        // this request has been redirected once or more. The registered
+        // listeners will received the following properties:
+        //
+        //   - `addonId`: the add-on ID that redirected the request, if any.
+        //   - `firstUrl`: the first URL loaded for the request, before any
+        //      redirects.
+        //   - `lastUrl`: the last URL loaded for the request, after one or
+        //      more redirects.
+        onRedirected: new ExtensionCommon.EventManager({
+          context,
+          name: "addonsSearchExperiment.onRedirected",
+          register: (fire, filter) => {
+            const stopListener = (event) => {
+              if (event.type != "stop") {
+                return;
+              }
+
+              const wrapper = event.currentTarget;
+              const { channel } = wrapper;
+
+              // When we detected a redirect, we read the request property,
+              // hoping to find an add-on ID corresponding to the add-on that
+              // initiated the redirect. It might not return anything when the
+              // redirect is a search server-side redirect but it can also be
+              // caused by an error.
+              let addonId;
+              try {
+                addonId = channel
+                  ?.QueryInterface(Ci.nsIPropertyBag)
+                  ?.getProperty("redirectedByExtension");
+              } catch (err) {
+                Cu.reportError(err);
+              }
+
+              let redirectChain;
+              try {
+                redirectChain = channel?.loadInfo?.redirectChain.map(
+                  (entry) => entry.principal.spec
+                );
+              } catch (err) {
+                Cu.reportError(err);
+              }
+
+              fire.sync({
+                addonId,
+                firstUrl: redirectChain && redirectChain[0],
```

Comment:
`firstUrl` is not necessarily the matched URL, it could be a non-matching URL that redirects to a URL that is matched by the `onBeforeRedirect` listener. Consequently, the logic in `background.js` will not be able to find the addonId associated with the request.

You should probably do the mapping of the matched URL to addonIds in the `onBeforeRedirect` listener instead of the background script (or at least pass the matched URL to the event listener).

Reply:
heh correct, so we probably don't care about the first url of the chain, we only care about the first matched url.